### PR TITLE
fix(headers): add limit to maximum headers that should be parsed

### LINKED ISSUE #256: DOS vulnerability by sending a request with many many headers
We currently only check for the size of a single header field rather than all headers combined, this means a malicious client can send a request with thousands of headers and quickly consume a lot of memory on the server before the user of hyper can even access the request to reject it.

### PR DESCRIPTION
**Pull request by seanmonstar**
Closes #256

### CODE REVIEW

File: src/header/mod.rs [hunk 1]
```diff
@@ -118,6 +118,8 @@ pub struct Headers {
     data: HashMap<HeaderName, MuCell<Item>>
 }
 
+const MAX_HEADERS_COUNT: u8 = 100;
```
Comment (reem) [inline src/header/mod.rs]: This feels a bit low. How about 256?

Comment (s-panferov) [inline src/header/mod.rs]: Maybe it's better to restrict the total size instead of the count?

Comment (seanmonstar) [inline src/header/mod.rs]: I noticed other parsers did indeed restrict the total size. I thought that would be complicated, but I just thought of a way.

Comment (s-panferov) [inline src/header/mod.rs]: We need to restrict the total size because actually it's possible to consume a lot of memory just by *one* header, e.g. Cookie.

Comment (seanmonstar) [inline src/header/mod.rs]: Well, that is already checked in `read_header`.

**Comment by seanmonstar** [comment]
Updated to count characters instead of headers.

**Comment by s-panferov** [comment]
:+1: Can we make it configurable somehow in the future?

File: src/header/mod.rs [hunk 2]
```diff
@@ -118,6 +118,10 @@ pub struct Headers {
     data: HashMap<HeaderName, MuCell<Item>>
 }
 
+// To prevent DOS from a server sending a never ending header.
+// The value was copied from curl.
+const MAX_HEADERS_COUNT: u32 = 100 * 1024;
```
Comment (s-panferov) [inline src/header/mod.rs]: Maybe it's better to rename it? How about MAX_HEADERS_LENGTH?

**Comment by seanmonstar** [comment]
Hm, I guess it could be configurable by passing it as an argument to `from_raw`...

File: src/header/mod.rs (hunk 2 above)
Comment (reem) [inline src/header/mod.rs]: Agreed.

**Pull request merged**